### PR TITLE
PWA functionality including offline mode (featured flagged)

### PR DESCRIPTION
**Pull request by microbit-robert**
Notable behaviours:
- Fallback to English when changing language offline (when the selected language assets aren’t cached).
- Shows image placeholders for uncached images when offline.
- The app will automatically trigger a page reload when it detects the service worker has been updated. The service worker is updated when any of the build assets are updated.

There are two levels of feature gating:
- It's only enabled for microbit-foundation builds to avoid others shipping it accidentally. You can change FEATURE_PWA in the build to change this. We're open to a better system of build-time feature flags if there's interest (at least allowing existing env vars to override would make sense).
- There's a client side feature flag. You have to set this via `localStorage.setItem("flags", "pwa")` for now if you want to test an actual PWA install. We'll default that flag on to release to the beta after building more confidence internally.

### CODE REVIEW

**Comment by microbit-matt-hillsdon** [comment]
> all available languages are precached

Does that mean everyone pays the cost of every language now? Before interactivity or after/in the background?

File: vite.config.ts
Comment (microbit-matt-hillsdon) [inline vite.config.ts:66]: Only worth addressing after other issues: we've parameterised the project name and dataset elsewhere so should do that here too (and below).

**Comment by microbit-robert** [comment]
> > all available languages are precached
> 
> Does that mean everyone pays the cost of every language now? Before interactivity or after/in the background?

For this version, yes. We need to be more specific in what we precache. Maybe nothing language specific and save that for runtime. It does feel like the plugin might not suit our needs. We'll either end up creating just the service worker from scratch, or ditch the plugin and do everything from scratch.

File: src/simulator/Simulator.tsx
Comment (microbit-robert) [inline src/simulator/Simulator.tsx:55]: TODO: Revert before merge.

File: src/index.tsx
Comment (microbit-robert) [inline src/index.tsx:16]: The best or most efficient way? It might be the only way.

**Comment by microbit-robert** [comment]
Toast shown when changing language while offline and language assets are not already cached. Text is a working placeholder and needs discussion.

image

File: .github/workflows/build.yml
Comment (microbit-robert) [inline .github/workflows/build.yml:28]: Let's confirm what the default behaviour should be.

File: src/flags.ts
Comment (microbit-robert) [inline src/flags.ts:70]: This needs more thought/discussion. If you install the PWA, close it, then reopen it, you'll lose the flag and the ability to add the flag. In the meantime, this change will make the review branch a bit more useful for testing.

File: vite.config.ts
Comment (microbit-robert) [inline vite.config.ts:48]: Hmmm. This works, but feels fragile.

File: src/language-server/client.ts
Comment (microbit-matt-hillsdon) [reply src/language-server/client.ts:195]: See https://github.com/microbit-foundation/pyright/pull/34

Comment (microbit-robert) [reply src/language-server/client.ts]: We now only precache the fallback language for pyright using the build of the PR above [d45eeae](https://github.com/microbit-foundation/python-editor-v3/pull/1170/commits/d45eeae52e485ca75433c825fdb5e199402aa13d)

**Comment by microbit-matt-hillsdon** [comment]
image

This feels a bit odd when in PWA mode as there's no obvious affordance to reload the page. Ideally we'd find a more neutral phrasing.

File: src/index.tsx
Comment (microbit-matt-hillsdon) [inline src/index.tsx:52]: Do we need an `else` here similar to the sim?
Scenario to test: use flag, update app, then don't use flag and should see latest

File: vite.config.ts
Comment (microbit-matt-hillsdon) [reply vite.config.ts:48]: I think we should assert that this changes html.

File: vite.config.ts
Comment (microbit-matt-hillsdon) [inline vite.config.ts:78]: Could baseUrl or a similar option partition the service worker caches for beta/v3 ?

Comment (microbit-robert) [reply vite.config.ts:78]: Maybe workbox.cacheId will work (main/beta or prod/staging)
https://developer.chrome.com/docs/workbox/modules/workbox-build#type-GeneratePartial

Comment (microbit-matt-hillsdon) [reply vite.config.ts:85]: Trying that now...

File: .github/workflows/build.yml
Comment (microbit-matt-hillsdon) [reply .github/workflows/build.yml]: I've flipped it to enable only for Foundation builds so that folks (especially students) don't get bitten by the SW when doing more experimental work.

File: vite.config.ts
Comment (microbit-matt-hillsdon) [reply vite.config.ts:93]: This looks good. We had to explicitly name the non-pre-cache caches.

File: src/index.tsx
Comment (microbit-matt-hillsdon) [reply src/index.tsx:61]: 61213d240c8931b5636bbfb739fcd7f3fa74cda8

**Review by microbit-matt-hillsdon — APPROVED** [review]

**Comment by microbit-matt-hillsdon** [comment]
I'll merge this tomorrow when I'm more available for any fallout.

**Comment by microbit-matt-hillsdon** [comment]
This isn't working in practice as we used scoped service workers but the scope is /v/beta/ but typically access is at /v/beta for staging/production. Ideally we wouldn't redirect to the less pretty URLs.

We could register against a slashless scope by adding a header. It's a prefix match so that's a bit awkward, e.g. /v/3 would also match /v/31 and /v/beta would match /v/betamax. The former seems worth some consideration. Alternatively we just have one and accept the extra reload when switching between versions.

**Comment by microbit-matt-hillsdon** [comment]
> We could register against a slashless scope by adding a header. It's a prefix match so that's a bit awkward

We did this and it now seems to be working. Still behind the flag you need to set in localStorage.

**Pull request merged**
